influxdb3_wal/src/snapshot_tracker.rs

```diff
@@ -56,43 +57,60 @@ impl SnapshotTracker {
+            ">>> wal periods and snapshots"
+        );
+
+        if !self.should_run_snapshot(force_snapshot) {
             return None;
         }
 
         // if the number of wal periods is >= 3x the snapshot size, snapshot everything up to, but
         // not including, the last period:
-        if self.wal_periods.len() >= 3 * self.snapshot_size {
-            let n_periods_to_take = self.wal_periods.len() - 1;
-            let wal_periods: Vec<WalPeriod> =
-                self.wal_periods.drain(0..n_periods_to_take).collect();
-            let max_time = wal_periods
-                .iter()
-                .map(|period| period.max_time)
-                .max()
-                .unwrap();
-            let t = max_time - (max_time.get() % self.gen1_duration.as_nanos())
-                + self.gen1_duration.as_nanos();
-            let last_wal_sequence_number = wal_periods.last().unwrap().wal_file_number;
-
-            let snapshot_details = SnapshotDetails {
-                snapshot_sequence_number: self.increment_snapshot_sequence_number(),
-                end_time_marker: t.get(),
-                last_wal_sequence_number,
-            };
-
-            return Some(SnapshotInfo {
-                snapshot_details,
-                wal_periods,
-            });
+        let wal_periods_3_times_snapshot_size = self.wal_periods.len() >= 3 * self.snapshot_size;
+        if force_snapshot || wal_periods_3_times_snapshot_size {
+            info!(
+                ?force_snapshot,
+                ?wal_periods_3_times_snapshot_size,
+                "snapshotting all"
+            );
+            return self.snapshot_all();
         }
 
-        let t = self.wal_periods.last().unwrap().max_time;
+        info!(
+            ?force_snapshot,
+            ?wal_periods_3_times_snapshot_size,
+            "snapshotting all before last wal period (using last wal period time)"
+        );
+        // uses the last wal period's time and snapshots everything before that, leaving just
+        // the last wal period, this will cater for all the writes that came in order
+        self.snapshot_in_order_wal_periods()
+    }
+
+    fn should_run_snapshot(&mut self, force_snapshot: bool) -> bool {
+        // wal buffer can be empty but wal periods shouldn't be
+        if self.wal_periods.is_empty() {
+            return false;
+        }
+
+        // if force_snapshot is set, we don't need to check wal periods len or num periods we
+        // snapshot immediately
+        if !force_snapshot && self.wal_periods.len() < self.number_of_periods_to_snapshot_after() {
+            return false;
+        }
+
+        true
+    }
+
+    pub(crate) fn snapshot_in_order_wal_periods(&mut self) -> Option<SnapshotDetails> {
+        let t = self.wal_periods.last()?.max_time;
         // round the last timestamp down to the gen1_duration
         let t = t - (t.get() % self.gen1_duration.as_nanos());
```

Comment:
This is the part that end ups leaving behind the older 300 periods. If max time is around now, and we round down from that with the gen1 duration (10m), we end up with a time 10 minutes in the past. Just a few lines down from this you see where it cycles through the WAL periods, taking all while its max time is less than this. So the logic is still the same, which is good 😄